### PR TITLE
Equal height row component responsive updates

### PR DESCRIPTION
**Pull request by pastelcyborg**
## Done

- Add new wrap variant to support new pattern designs from Figma
- Add new examples and docs for wrap variant
- Drive-by: Code cleanup and formatting

Fixes [WD-13923](https://warthogs.atlassian.net/browse/WD-13923)

## QA

- Open [docs](https://vanilla-framework-5276.demos.haus/docs/patterns/equal-height-row) and [demo](https://vanilla-framework-5276.demos.haus/docs/examples/patterns/equal-height-row/combined?theme=light) for new wrap variant
- Confirm docs accurately represent new/old variants
- Confirm new wrap variant matches designs and expected functionality, particularly at medium screen sizes

### Check if PR is ready for release

If this PR contains Vanilla SCSS code changes, it should contain the following changes to make sure it's ready for the release:

- [X] PR should have one of the following labels to automatically categorise it in release notes:
  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
- [X] Vanilla version in `package.json` should be updated relative to the [most recent release](https://github.com/canonical/vanilla-framework/releases/latest), following semver convention:
  - if CSS class names are not changed it can be bugfix release (x.x.**X**)
  - if CSS class names are changed/added/removed it should be minor version (x.**X**.0)
  - see the [wiki for more details](https://github.com/canonical/vanilla-framework/wiki/Release-process#pre-release-tasks)
- [X] Any changes to component class names (new patterns, variants, removed or added features) should be listed on the [what's new page](https://github.com/canonical/vanilla-framework/blob/main/releases.yml).

[WD-13923]: https://warthogs.atlassian.net/browse/WD-13923?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: templates/docs/patterns/equal-height-row.md [hunk 1]
```diff
@@ -4,72 +4,110 @@ context:
...
-  </tbody>
-</table>
-
-This component is an extension of the [grid component](/docs/patterns/grid). There are several usage variations detailed in the examples below.
+The equal height row component aims to provide consistent alignment for grid
+items within a row format. This is achieved using the CSS `subgrid` feature
+which allows column grids to share the same layout used in the parent row grid.
+
+In addition to enforcing column grid item alignment within each row, default
+responsive behavior is shown below:
```
Comment (jmuzina) [inline templates/docs/patterns/equal-height-row.md:12]: I think MDspell is failing this PR due to American spelling of "behavior". Try "behaviour", here and elsewhere

```suggestion
responsive behaviour is shown below:
```

File: scss/_patterns_equal-height-row.scss [hunk 2]
```diff
@@ -18,18 +20,28 @@
       margin-top: -1px;
       position: relative;
 
+      // If first row item contains an image container, apply top margin to
+      // match top margin from deprecated media container component
+      .p-equal-height-row__item:first-child .p-image-container {
```
Comment (jmuzina) [inline scss/_patterns_equal-height-row.scss:25]: I don't think targeting class names from another component is ideal. Could we achieve this some other way? Maybe a modifier class?

File: templates/docs/examples/patterns/equal-height-row/4-items-per-column-wrap.html [hunk 3]
```diff
@@ -0,0 +1,139 @@
+{% extends "_layouts/examples.html" %}
+{% block title %}Equal height row / Four items per column and full-screen divider, wrap{% endblock %}
+
+{% block standalone_css %}patterns_equal-height-row{% endblock %}
+
+{% block content %}
+<div class="p-equal-height-row--wrap has-divider-3">
+  <div class="p-equal-height-row__col">
+    <div class="p-equal-height-row__item">
+      <div class="p-image-container p-image-container--2-3 is-highlighted">
```
Comment (jmuzina) [inline templates/docs/examples/patterns/equal-height-row/4-items-per-column-wrap.html:10]: `p-image-container--2-3` has all the styles from `p-image-container` because of the weird way we do BEM. So all that is needed is:

```suggestion
      <div class="p-image-container--2-3 is-highlighted">
```

File: templates/docs/examples/patterns/equal-height-row/4-items-per-column-wrap.html [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        <p>
+          Learn more about Ubuntu security&nbsp;›
+        </p>
+      </a>
+    </div>
+  </div>
+
+  <div class="p-equal-height-row__col">
+    <div class="p-equal-height-row__item">
+      <div class="p-image-container p-image-container--2-3 is-highlighted">
```
Comment (jmuzina) [inline templates/docs/examples/patterns/equal-height-row/4-items-per-column-wrap.html:43]:
```suggestion
      <div class="p-image-container--2-3 is-highlighted">
```

File: templates/docs/examples/patterns/equal-height-row/4-items-per-column-wrap.html [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        <p>
+          Learn more about fleet management&nbsp;›
+        </p>
+      </a>
+    </div>
+  </div>
+
+  <div class="p-equal-height-row__col">
+    <div class="p-equal-height-row__item">
+      <div class="p-image-container p-image-container--2-3 is-highlighted">
```
Comment (jmuzina) [inline templates/docs/examples/patterns/equal-height-row/4-items-per-column-wrap.html:76]:
```suggestion
      <div class="p-image-container--2-3 is-highlighted">
```

File: templates/docs/examples/patterns/equal-height-row/4-items-per-column-wrap.html [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        <p>
+          Learn more about confinement&nbsp;›
+        </p>
+      </a>
+    </div>
+  </div>
+
+  <div class="p-equal-height-row__col">
+    <div class="p-equal-height-row__item">
+      <div class="p-image-container p-image-container--2-3 is-highlighted">
```
Comment (jmuzina) [inline templates/docs/examples/patterns/equal-height-row/4-items-per-column-wrap.html:109]:
```suggestion
      <div class="p-image-container--2-3 is-highlighted">
```

File: templates/docs/examples/patterns/equal-height-row/4-items-per-column.html [hunk 7]
```diff
@@ -4,97 +4,136 @@
...
 {% block content %}
-
 <div class="p-equal-height-row has-divider-3">
   <div class="p-equal-height-row__col">
     <div class="p-equal-height-row__item">
-      <div class="p-media-container">
-        <img width="320"
-          src="https://assets.ubuntu.com/v1/c4176cd9-Kernelt%20industries-80%201.jpg"
-          alt="" />
+      <div class="p-image-container p-image-container--2-3 is-highlighted">
```
Comment (jmuzina) [inline templates/docs/examples/patterns/equal-height-row/4-items-per-column.html:10]:
```suggestion
      <div class="p-image-container--2-3 is-highlighted">
```

File: templates/docs/examples/patterns/equal-height-row/4-items-per-column.html [hunk 8]
```diff
@@ -4,97 +4,136 @@
...
     </div>
   </div>
+
   <div class="p-equal-height-row__col">
     <div class="p-equal-height-row__item">
-      <div class="p-media-container">
-        <img width="320"
-          src="https://assets.ubuntu.com/v1/1624c652-Kernelt%20industries-80%201-1.jpg"
-          alt="" />
+      <div class="p-image-container p-image-container--2-3 is-highlighted">
```
Comment (jmuzina) [inline templates/docs/examples/patterns/equal-height-row/4-items-per-column.html:43]:
```suggestion
      <div class="p-image-container--2-3 is-highlighted">
```

File: templates/docs/examples/patterns/equal-height-row/4-items-per-column.html [hunk 9]
```diff
@@ -4,97 +4,136 @@
...
     </div>
   </div>
+
   <div class="p-equal-height-row__col">
     <div class="p-equal-height-row__item">
-      <div class="p-media-container">
-        <img width="320"
-          src="https://assets.ubuntu.com/v1/eb9824bf-Kernelt%20industries-80%201-2.jpg"
-          alt="" />
+      <div class="p-image-container p-image-container--2-3 is-highlighted">
```
Comment (jmuzina) [inline templates/docs/examples/patterns/equal-height-row/4-items-per-column.html:76]:
```suggestion
      <div class="p-image-container--2-3 is-highlighted">
```

File: templates/docs/examples/patterns/equal-height-row/4-items-per-column.html [hunk 10]
```diff
@@ -4,97 +4,136 @@
...
     </div>
   </div>
+
   <div class="p-equal-height-row__col">
     <div class="p-equal-height-row__item">
-      <div class="p-media-container">
-        <img width="320"
-          src="https://assets.ubuntu.com/v1/000a7695-Kernelt%20industries-80%201-3.jpg"
-          alt="" />
+      <div class="p-image-container p-image-container--2-3 is-highlighted">
```
Comment (jmuzina) [inline templates/docs/examples/patterns/equal-height-row/4-items-per-column.html:109]:
```suggestion
      <div class="p-image-container--2-3 is-highlighted">
```

**Review by jmuzina — CHANGES_REQUESTED** [review]
Left some comments about fixing typos in documentation and using terser image container classes

**Comment by jmuzina** [comment]
Should this divider on medium be full-width or just on the right side?
Screenshot 2024-08-07 at 2 23 32 PM

**Comment by bartaz** [comment]
> 1. Change grid breakpoints to match `u-hide` breakpoints, which would alter grid functionality across the entire library
> 2. Change `u-hide` breakpoints, which would alter `u-hide` functionality across the entire library
> 
> @bartaz Thoughts?

We used to have similar issues in the past. Seems something slipped by the fixes so far.
I'd have to look into details.

But overall, yes, we will need to adjust either one of these or both, so that they cover the ranges in the same way.
It used to be tricky with `min-width` `max-width` media queries, but now with new `width <` `width >=` etc we should be able to exactly match those.

I'd say it doesn't matter for this PR (I guess if we fix it properly separately, it will automatically fix here), so we should address it separately.

**Comment by jmuzina** [comment]
This issue is unrelated to breakpoints - it's much simpler. Looks like the hide utility needs to be added to standalone imports here

**Comment by jmuzina** [comment]
> I've resolved 1 and 2. 3 is still a problem - this screenshot was taken at 1035.72px. Will continue to explore a solution.

@pastelcyborg I don't think the third one could be solved within the scope of this PR, maybe it should be filed as a separate issue to be solved later

File: scss/_patterns_equal-height-row.scss (hunk 2 above)
Comment (bartaz) [reply scss/_patterns_equal-height-row.scss:25]: I think this relates to another issue that we have.

In the initial design for horizontal rules (`p-rule` component and its vatiants) rules did not have a bottom margin. It was OK when they have been followed by text content, but images required a top margin (and so the now deprecated `p-media-container` was born to add this margin).

But after struggling it for some time Lyubo decided that it's OK for the text to be further away from the horizontal lines, so that horizontal lines can have bottom margin, and images don't need to have the top margin anymore.

I guess the issue is, that currently only class-less base `<hr>` has bottom margin, but `p-rule` doesn't.

There actually is a PR by Pete to address that: https://github.com/canonical/vanilla-framework/pull/5241 but it kind of got stuck in review limbo.

So, to summarise. The equal heights should not be doing anything to add any margins to images. If there are lines added by the pattern, the lines themselves should have a bottom margin to move everything below them by 0.5rem.
If the lines are `p-rule` component, it's the rule component that needs to be updated to have bottom margin.

I guess we can achieve that by finishing the Pete's PR.

**Comment by lyubomir-popov** [comment]
A couple of quick thoughts, but I think we should spend some time on this in the standup:
- I think the 3 col version for medium would be great to add. For special cases, with 3 items for example. Of course, that would be appropriate only for short texts
- would be good to demonstrate the new aaspect ratio change classes, where  for example the 2:3 images become huge

File: scss/_patterns_equal-height-row.scss [hunk 11]
```diff
@@ -15,21 +17,24 @@
...
         }
 
-        // for medium screen, position the first column item on the left of the grid
+        // At medium size, position the first column item on the left of the
+        // grid
         .p-equal-height-row__item:first-child {
-          grid-row: span 100; // this needs to be sufficiently large so remaining column items won't get stretched
+          // This needs to be sufficiently large so remaining column items won't
+          // get stretched
+          grid-row: span 100;
```
Comment (jmuzina) [inline scss/_patterns_equal-height-row.scss:37]: Nit: There's a bit of weirdness here with single-sentence comments being split into multiple single-line comments. I'm guessing a formatter did this.

If these sentences have to be multiple lines for readability they should use multi-line comments

File: templates/docs/examples/patterns/equal-height-row/4-items-per-column-responsive.html [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    <p class="p-equal-height-row__item">
+      Imagine a world where every device is trustworthy. We designed every
+      aspect of Ubuntu Core to create the most secure embedded Linux ever, with
+      a 10 year LTS commitment.
+    </p>
+
+    <div class="p-equal-height-row__item">
+      <a href="#">
+        <p>
+          Learn more about Ubuntu security&nbsp;›
```
Comment (jmuzina) [inline templates/docs/examples/patterns/equal-height-row/4-items-per-column-responsive.html:35]: Nit: use HTML entity for the right chevron here and anywhere else it is found
```suggestion
          Learn more about Ubuntu security&nbsp;&rsaquo;
```

File: scss/standalone/patterns_equal-height-row.scss [hunk 13]
```diff
@@ -4,7 +4,10 @@
 // dependencies needed for examples
 @include vf-p-buttons;
 @include vf-p-media-container;
+@include vf-p-image;
 @include vf-p-headings;
```
Comment (jmuzina) [inline scss/standalone/patterns_equal-height-row.scss:8]: Looks like standalone is missing the import for muted headings
```suggestion
@include vf-p-headings;
@include vf-p-muted-heading;
```
Non-standalone vs standalone:
Screenshot 2024-08-22 at 5 29 00 PM
Screenshot 2024-08-22 at 5 29 03 PM

File: templates/docs/examples/patterns/equal-height-row/3-column-row.html
Comment (jmuzina) [inline templates/docs/examples/patterns/equal-height-row/3-column-row.html:1]: It would be nice to add `redirects.yaml` redirects from these three deleted examples to their responsive versions, in case anybody has any links to these examples lying around.

File: releases.yml [hunk 14]
```diff
@@ -54,6 +54,10 @@
       url: /docs/patterns/tiered-list
       status: New
       notes: We've introduced a new tiered list pattern.
+    - component: Equal Height Row
+      url: /docs/patterns/equal-height-row#wrap-variant
+      status: New
+      notes: We've added a new wrap variant to the equal height row component.
```
Comment (jmuzina) [inline releases.yml:60]: This has been added to `4.15.0` release notes section, but it should be added to `4.16.0`.

Screenshot 2024-08-22 at 5 33 29 PM

**Review by jmuzina — CHANGES_REQUESTED** [review]

**Comment by jmuzina** [comment]
If it's not too much to ask, can we also add a SCSS class reference table to the docs page for this component? It's starting to have a lot of different classes and it would be nice to add it.

[Image](https://vanillaframework.io/docs/patterns/images#class-reference) is a good example of one of these. We don't use this feature often but we should get in the habit of doing it.

**Review by jmuzina — APPROVED** [review]
